packages/material-ui/src/Menu/Menu.js

```diff
@@ -115,37 +168,119 @@ const Menu = React.forwardRef(function Menu(props, ref) {
     }
   });
 
+  const handleSetOpenSubMenuIndex = (value) => {
+    if (value === null) {
+      if (contentAnchorRef.current.parentElement) {
+        contentAnchorRef.current.parentElement.children[openSubMenuIndex].focus();
+      }
+    }
+    setOpenSubMenuIndex(value);
+  };
+
   const items = React.Children.map(children, (child, index) => {
+    if (!React.isValidElement(child)) {
+      return;
+    }
+
+    const { subMenu, onMouseMove: onMouseMoveChildProp } = child.props;
+    const { anchorEl } = other;
+
+    const hasSubMenu = Boolean(subMenu);
+    const parentMenuOpen = Boolean(anchorEl);
+
+    let additionalPropsAdded = false;
+    const additionalProps = {};
+
+    // This is the original purpose of this React.Children.map and is basically unchanged.
     if (index === activeItemIndex) {
-      return React.cloneElement(child, {
+      additionalPropsAdded = true;
+
+      Object.assign(additionalProps, {
         ref: (instance) => {
           contentAnchorRef.current = instance;
           setRef(child.ref, instance);
         },
       });
     }
 
+    // If the current Menu item in this map has a subMenu,
+    // we need the parent Menu to orchestrate its subMenu
+    if (hasSubMenu && parentMenuOpen) {
+      additionalPropsAdded = true;
+
+      const handleArrowRightKeydown = (event) => {
+        if (event.key === 'ArrowRight') {
+          event.preventDefault();
+          setOpenSubMenuIndex(index);
+        }
+      };
+
+      Object.assign(additionalProps, {
+        handleArrowRightKeydown,
+        openSubMenu: index === openSubMenuIndex && !entering,
+        setParentOpenSubMenuIndex: handleSetOpenSubMenuIndex,
+      });
+    }
+
+    // If there are ANY children with subMenus, then ALL
+    // of the children need to know how to close any open subMenus
+    // and reset the state that controls which subMenu is open.
+    if (atLeastOneSubMenu) {
+      additionalPropsAdded = true;
+
+      Object.assign(additionalProps, {
+        onMouseMove: (e) => {
+          setOpenSubMenuIndex(index);
+          if (onMouseMoveChildProp) {
+            onMouseMoveChildProp(e);
+          }
+        },
+        handleParentClose: handleOnClose,
+      });
+    }
+
+    // Using a semaphore instead of inspecting addtionalProps
+    // directly to avoid performance hits at scale. Might be
+    // fine to just do Object.keys(additionalProps).length > 0,
+    // but that seems like iterations we can avoid.
+    if (additionalPropsAdded) {
+      // eslint-disable-next-line consistent-return
```

Comment:
Please address this lint rule.

Reply:
I tried to find a way much earlier in this process but failed to find a way. I'll take another look, but it seems like it's simply due to the fact we either return the child or a clone of it, which is the entire reason for this code block.

Reply:
It's somewhat indicative that the function might be too complicated. If you don't know where it returns when you actually wrote it how would somebody be able to reason about the Menu component with no knowledge of this PR?

Reply:
It only returns from one of the two places at the bottom of the block. Either it returns in the `if..` block, if the condition is met, or it returns via the final return. Maybe just moving it to an `if..else` would make it happier, but I'm pretty sure I tried that.

I was chasing little things all day with this to get it to a happy place for the CI. I should be pushing that either shortly, or tomorrow sometime, and I'll start circling back around on stuff like this.

Reply:
> I was chasing little things all day with this to get it to a happy place for the CI.

You should be able to reproduce this locally. Is `yarn lint` not working for you?

Reply:
I'm passing all the linting tests. I also created a new component (SubMenu) and had to get that to a place where it had at least a test stub, and was able to generate api docs, etc. There were a bunch of little things I had started and not finished when I ran into the focus regression after rebasing. I'm working on this more today, so I'll have updates coming in throughout the day.

Reply:
Fixed this and pushed.